Web Inspector: Add a setting to turn off dimming unrendered nodes

patrickangle opened the pull request:
#### bbe97632cc4eea1ce837b64335495e6b1ab7ec3a
<pre>
Web Inspector: Add a setting to turn off dimming unrendered nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=244361">https://bugs.webkit.org/show_bug.cgi?id=244361</a>
rdar://99155655

Reviewed by Devin Rousso.

For some use cases it can still be useful to get full-color non-rendered elements in the DOM tree, particularly when
looking through and modifying elements inside the &lt;head&gt; of the document, so add a setting to turn the enable/disable
the dimming (defaults to enabled).

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView):
(WI.DOMTreeContentView.prototype.closed):
(WI.DOMTreeContentView.prototype._updateDOMTreeDeemphasizesNodeThatAreNotRendered):
(WI.DOMTreeContentView.prototype._handleDOMTreeDeemphasizesNodesThatAreNotRenderedChanged):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeOutline.css:
(.content-view.dom-tree.deemphasize-unrendered .tree-outline.dom li:not(.rendered) &gt; span &gt; *:not(.html-comment)):
(.content-view.dom-tree.deemphasize-unrendered .tree-outline.dom li:not(.rendered, .selected) &gt; span &gt; *:not(.html-comment),):
(.content-view.dom-tree .tree-outline.dom li:not(.rendered) &gt; span &gt; *:not(.html-comment)): Deleted.
(.content-view.dom-tree .tree-outline.dom li:not(.rendered, .selected) &gt; span &gt; *:not(.html-comment),): Deleted.
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createElementsSettingsView):

Canonical link: <a href="https://commits.webkit.org/254022@main">https://commits.webkit.org/254022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4512d57ee6ac46a85ad8232756512bceb14e498c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87805 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31900 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/18531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/96992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/151315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30261 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/91736 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93423 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/79902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/79397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/18531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27935 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/18531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27916 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/18531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2825 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29604 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29498 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/18531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -279,6 +279,12 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
 
         let elementsSettingsView = new WI.SettingsView("elements", WI.UIString("Elements"));
 
+        if (InspectorBackend.hasCommand("DOM.setInspectModeEnabled", "showRulers")) {
```
oops?  this should probably be something like
```
// COMPATIBILITY (macOS 13.0, iOS 16.0): CSS.LayoutFlag.Rendered did not exist yet.
if (InspectorBackend.Enum.CSS?.LayoutFlag?.Rendered) {
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DOMTreeElement.js:
```diff
@@ -2167,17 +2169,28 @@ WI.DOMTreeElement = class DOMTreeElement extends WI.TreeElement
+    _updateUnrenderedNodeDimming()
     {
-        this.listItemElement?.classList.toggle("rendered", this.representedObject.layoutFlags.includes(WI.DOMNode.LayoutFlag.Rendered));
+        let renderedAndDimmingEnabled = !WI.settings.domTreeDimsUnrenderedNodes.value
```
I think rather than have every `WI.DOMTreeElement` listen for this change, maybe we have the `WI.DOMTreeOutline` (or even the `WI.DOMTreeContentView`) listen for it and set a top-level CSS property that is then used inside
```
.content-view.dom-tree.indicate-rendered .tree-outline.dom li:not(.rendered) > span > *:not(.html-comment)
                      ^^^^^^^^^^^^^^^^^^
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
```diff
@@ -279,6 +279,12 @@ WI.SettingsTabContentView = class SettingsTabContentView extends WI.TabContentVi
         let elementsSettingsView = new WI.SettingsView("elements", WI.UIString("Elements"));
 
+        if (InspectorBackend.hasCommand("DOM.setInspectModeEnabled", "showRulers")) {
+            elementsSettingsView.addSetting(WI.UIString("DOM Tree:"), WI.settings.domTreeDimsUnrenderedNodes, WI.UIString("Dim unrendered nodes"));
```
I think maybe the text could be a bit more verbose to help developers understand.  Maybe something like

> De-emphasize nodes that are not rendered.

or 

> Differentiate nodes that will never paint.

(and I'd probably rename the `WI.Setting` to match)

patrickangle committed:
Web Inspector: Add a setting to turn off dimming unrendered nodes
https://bugs.webkit.org/show_bug.cgi?id=244361
rdar://99155655

Reviewed by Devin Rousso.

For some use cases it can still be useful to get full-color non-rendered elements in the DOM tree, particularly when
looking through and modifying elements inside the <head> of the document, so add a setting to turn the enable/disable
the dimming (defaults to enabled).

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/DOMTreeContentView.js:
(WI.DOMTreeContentView):
(WI.DOMTreeContentView.prototype.closed):
(WI.DOMTreeContentView.prototype._updateDOMTreeDeemphasizesNodeThatAreNotRendered):
(WI.DOMTreeContentView.prototype._handleDOMTreeDeemphasizesNodesThatAreNotRenderedChanged):
* Source/WebInspectorUI/UserInterface/Views/DOMTreeOutline.css:
(.content-view.dom-tree.deemphasize-unrendered .tree-outline.dom li:not(.rendered) > span > *:not(.html-comment)):
(.content-view.dom-tree.deemphasize-unrendered .tree-outline.dom li:not(.rendered, .selected) > span > *:not(.html-comment),):
(.content-view.dom-tree .tree-outline.dom li:not(.rendered) > span > *:not(.html-comment)): Deleted.
(.content-view.dom-tree .tree-outline.dom li:not(.rendered, .selected) > span > *:not(.html-comment),): Deleted.
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createElementsSettingsView):

Canonical link: https://commits.webkit.org/254022@main
